### DISCUSSION TITLE
Classes as Dependencies: New instance runs in thread?

### DISCUSSION THREAD

**Discussion by conradogarciaberrotaran** [fastapi/fastapi#13064] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from typing import Annotated

from fastapi import Depends, FastAPI

app = FastAPI()


fake_items_db = [{"item_name": "Foo"}, {"item_name": "Bar"}, {"item_name": "Baz"}]


class CommonQueryParams:
    def __init__(self, q: str | None = None, skip: int = 0, limit: int = 100):
        self.q = q
        self.skip = skip
        self.limit = limit
@app.get("/items/")
async def read_items(commons: Annotated[CommonQueryParams, Depends()]):
    response = {}
    if commons.q:
        response.update({"q": commons.q})
    items = fake_items_db[commons.skip : commons.skip + commons.limit]
    response.update({"items": items})
    return response
```

### Description

Might be related to Uvicorn, but also related to FastaAPI's dependency injection. So I'm not completely sure.

Question is: Are we running the new instance of a class as dependency in a separate thread? if so, this might be an avoidable performance hit.

### Operating System

Linux

### Operating System Details

Doesnt apply

### FastAPI Version

latest

### Pydantic Version

latest

### Python Version

3.x

**Comment by Kludex** [comment, marked as answer]
It's FastAPI.

> Are we running the new instance of a class as dependency in a separate thread? 

Yes.

Read more about it on [FastAPI Tips](https://github.com/Kludex/fastapi-tips#9-your-dependencies-may-be-running-on-threads).

**Comment by conradogarciaberrotaran** [comment]
IMO, should be stated in the docs I linked above.

or change the example to
````python

class CommonQueryParams:
    async def __call__(self, q: str | None = None, skip: int = 0, limit: int = 100):
        self.q = q
        self.skip = skip
        self.limit = limit
````

**Reply by Kludex** [reply]
I agree something needs to change.

**Reply by conradogarciaberrotaran** [reply]
Just tried this case:

```python
class CommonQueryParams:
    async def __call__(self, q: str | None = None, skip: int = 0, limit: int = 100):
        self.q = q
        self.skip = skip
        self.limit = limit
```
and it's not working, the `__call__` method is not called.
This is what I tried:
```python

@router.get(
    "/",
)
async def home(query_params: Annotated[CommonQueryParams, Depends()]):
    ...
```
or 

```python

@router.get(
    "/",
)
async def home(query_params: CommonQueryParams = Depends()):
    ...
```
or

```python

@router.get(
    "/",
)
async def home(query_params: CommonQueryParams = Depends(CommonQueryParams)):
    ...
```
So options are:

1. Leave it as is and add a warning to the docs mentioning the potential performance hit (additionaly, we could add it on fastapi-tips as well)
2. Create an instance in the main event loop and potentially block the event loop
3. Make the async def __call__ work in the event loop.

What do you think?

**Reply by Kludex** [reply]
> Create an instance in the main event loop and potentially block the event loop

I didn't get this.

> Make the async def call work in the event loop.

I have this: https://github.com/Kludex/fastapi-dependency... Does it help?

**Comment by conradogarciaberrotaran** [comment]
> Create an instance in the main event loop and potentially block the event loop

Call the `__init__` in the main loop, it can potentially block the event loop if the user decides to create a slow `__init__` function.
If it's not slow, the overhead of creating a new thread just to instantiate a class it's not worth it.

> I have this: https://github.com/Kludex/fastapi-dependency... Does it help?

This would make it more obvious, but not fix the issue.

**Reply by Kludex** [reply]
How'd you know that you should run that in the event loop?

**Reply by conradogarciaberrotaran** [reply]
I wouldn't know what's best. But IMO, creating a thread just to instantiate a class is not worth it. I imagine most cases looks like this:

```python

class CommonQueryParams:
    def __init__(self, q: str | None = None, skip: int = 0, limit: int = 100):
        self.q = q
        self.skip = skip
        self.limit = limit
        
```

If somebody decided to do this:

```python

class CommonQueryParams:
    def __init__(self, q: str | None = None, skip: int = 0, limit: int = 100):
        self.q = q
        self.skip = skip
        sleep(10)
        self.limit = limit
```

Then, spinning up the new thread is definitely worth it.

For now, we don't have any way to decide.

**Reply by conradogarciaberrotaran** [reply]
For future reference, this is where it's happening:
https://github.com/fastapi/fastapi/blob/master/fastapi/dependencies/utils.py#L640
